sklearn/ensemble/_hist_gradient_boosting/_histogram_pool.py

```diff
@@ -0,0 +1,35 @@
+import numpy as np
+from .common import HISTOGRAM_DTYPE
+
+
+class HistogramsPool:
+    """Histograms pool to be used by the growers.
+
+    The pool allocates and returns histograms to the caller. When the `reset`
+    method is called, all the previously allocated histograms will be available
+    for the next grower to use. New histograms will be allocated when there are
+    no histograms left in the available_pool. HistogramsPool is used for memory
+    allocation/management only. The computation of the histograms is done in
+    HistogramBuilder.
```

Comment:
Maybe make it explicit that empirically, this strategy has proven much more memory efficient than allocating new histograms each time and relying on the Python GC to keep the overall python process memory low when fitting GBRT on datasets with many features and target classes.

Reply:
I believe this comment might be outdated considering that the memory boost mainly came from removing cycles